src/lib/auth/performAuth.js

```diff
@@ -0,0 +1,120 @@
+        response_type: "code",
+        client_id: CLIENT_ID,
+        state,
+        scope,
+        redirect_uri: AUTH_CALLBACK_URL,
+        code_challenge: await pkceChallengeFromVerifier(verifier),
+        code_challenge_method: "S256",
+    }).toString();
+};
+
+const DEFAULT_REDIRECT = withBasePath("/overview");
+
+export const performAuth = async (authorizationEndpoint, scope = "openid email") => {
+    window.location = await createAuthURL(authorizationEndpoint, scope);
+};
+
+const defaultAuthCodeCallback = async (dispatch, history, code, verifier) => {
+    const lastPath = localStorage.getItem(LS_BENTO_POST_AUTH_REDIRECT);
+    localStorage.removeItem(LS_BENTO_POST_AUTH_REDIRECT);
+
+    await dispatch(tokenHandoff(code, verifier));
+    history.replace(lastPath ?? DEFAULT_REDIRECT);
+    await dispatch(fetchUserDependentData(nop));
+};
+
+export const setLSNotSignedIn = () => {
+    localStorage.removeItem(LS_BENTO_WAS_SIGNED_IN);
+};
+
+export const useHandleCallback = (callbackPath, authCodeCallback = undefined) => {
+    const dispatch = useDispatch();
+    const history = useHistory();
+    const location = useLocation();
+    const oidcConfig = useSelector(state => state.openIdConfiguration.data);
+    const idTokenContents = useSelector(state => state.auth.idTokenContents);
+    const isAuthenticated = getIsAuthenticated(idTokenContents);
+
+    useEffect(() => {
+        // Ignore non-callback URLs
+        if (!location.pathname.startsWith(callbackPath)) return;
+
+        // End early if we don't have OpenID config (yet)
+        if (!oidcConfig) return;
+
+        // If we're already authenticated, don't try to reauthenticate
+        if (isAuthenticated) {
+            history.replace(DEFAULT_REDIRECT);
+            return;
+        }
+
+        const params = new URLSearchParams(window.location.search);
+
+        const error = params.get("error");
+        if (error) {
+            message.error(`Error encountered during sign-in: ${error}`);
+            console.error(error);
+            setLSNotSignedIn();
+            return;
+        }
+
+        const code = params.get("code");
+        if (!code) {
+            // No code, don't do anything
+            setLSNotSignedIn();
+            return;
+        }
+
+        const localState = localStorage.getItem(PKCE_LS_STATE);
+        localStorage.removeItem(PKCE_LS_STATE);
+
+        if (!localState) {
+            console.error("no local state");
+            setLSNotSignedIn();
+            return;
+        }
+
+        const paramState = params.get("state");
+        if (localState !== paramState) {
+            message.error("Error encountered during sign-in: state mismatch");
+            console.error("state mismatch");
+            setLSNotSignedIn();
+            return;
+        }
+
+        const verifier = localStorage.getItem(PKCE_LS_VERIFIER);
+        localStorage.removeItem(PKCE_LS_VERIFIER);
```

Comment:
Minor. We could wrap the few repetitions of `localStorage.getItem(key) ... localStorate.removeItem(key)` in a utils function named something like `popLsItem`

Reply:
good idea